src/lib.rs

```diff
@@ -0,0 +1,489 @@
+// Copyright (c) 2017 King's College London
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#[macro_use] extern crate lazy_static;
+#[macro_use] extern crate log;
+extern crate env_logger;
+
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::thread;
+use std::thread::JoinHandle;
+use std::cell::RefCell;
+
+
+/// Uniquely identifies a location in the user-program.
+///
+/// XXX Hard-coded to a simple program counter value.
+/// XXX This need to be user-specified, perhaps using a trait?
+type JitLocation = u32;
+
+/// The type used to count how many times a JitLocation has been interpreted.
+type HotCount = u32;
```

Comment:
My guess is that we don't want a `u32` here. It's way bigger than the hot count is, but not large enough to capture all possible executions of a bytecode. I suggest a `u16` as a safe long-term alternative: presumably, once the hot count is exceeded, we stop incrementing instances of this type?

Reply:
I'm ok with `u16`.

The hot count does actually keep incrementing until an executable trace appears, so that could be another concurrency bug.

Reply:
I think the hot count should stop incrementing at the point that tracing starts otherwise, whatever type we choose, overflow could in theory occur.

Reply:
Indeed.

Reply:
What do you mean by "not large enough to capture all possible executions of a bytecode" here actually?

Reply:
```rust
for 0..u32::max_size() + 1 {
    ...
}
```

Boom.

Reply:
Did you mean "large enough" instead of "not large enough" then? I understand we don't want overflows.

Reply:
No, "not large enough" was correct. The problem is that if you keep incrementing the counter even after a trace has started, you're going to overflow, then (maybe) hit the "limit" again (and again and again). Once the limit is hit, we *have* to stop incrementing.

Reply:
> My guess is that we don't want a u32 here. It's way bigger than the hot count is, but not large enough to capture all possible executions of a bytecode.

I don't understand this sentence (but I agree with what you are proposing nonetheless).

Reply:
Then let's go with the proposal ;)